rfcs/react-components/convergence/static-classes-on-slots.md

```diff
@@ -0,0 +1,90 @@
+
+History:
+
+- [PR #21206](https://github.com/microsoft/fluentui/pull/21206) - Initial submission
+
+## Summary
+
+This RFC proposes adding a static className to every slot, with a consistent and predictable naming format.
+
+## Problem statement
+
+It can be useful to have a static class name on the slot of a component, including by users of our library if they want to write static styles targeting a slot.
+
+We currently have a static class name `"fui-{ComponentName}"` (e.g. `"fui-Button"`) added to the root slot of every component, as described in issue [#19937](https://github.com/microsoft/fluentui/issues/19937). However, we don't currently have any formal specification of the format of the classNames for other slots, nor any conformance tests ensuring that they are added to every slot.
+
+Some of our components already have static classNames on slots (e.g. Checkbox's `indicator` slot). However, this patchwork addition of static classes makes it unpredictable for users of our library.
+
+One example internal usage would be to write a style such as this, which changes the Checkbox `indicator` slot's backgroundColor when hovering over the `root`:
+
+```ts
+const useStyles = makeStyles({
+  root: {
+    ':hover': {
+      '& .fui-Checkbox-indicator': {
+        backgroundColor: tokens.colorCompoundBrandBackgroundHover,
+      },
+    },
+  },
+});
+```
+
+## Detailed Design or Proposal
+
+### Slot Names
+
+The root slot already has a static className with the format `fui-{ComponentName}`, per [#19937](https://github.com/microsoft/fluentui/issues/19937). This RFC proposes that every other slot of a component should have a static className added, with the format `fui-{ComponentName}-{slot}`. The static className would be added to the start before any other classNames.
+
+For example, the HTML tree of Checkbox would look like:
+
+```html
+<span className="fui-Checkbox ...other classes...">
+  <span className="fui-Checkbox-indicator ...other classes..."> ... </span>
+  <input className="fui-Checkbox-input ...other classes..." />
+  <label className="fui-Checkbox-label ...other classes..."> ... </label>
+</span>
+```
+
+### Exports
+
+These classNames could be exported from each components in an object. This would replace the existing `{component}ClassName` export.
+
+Although these names could be automatically generated, there is value in having them be explicitly written out in the source code so it can be found via searching.
+
+```ts
+export const checkboxClassNames: ClassNames<'Checkbox', CheckboxSlots> = {
+  root: 'fui-Checkbox',
+  label: 'fui-Checkbox-label',
+  input: 'fui-Checkbox-input',
+  indicator: 'fui-Checkbox-indicator',
+};
+```
+
+With the helper type that ensures every slot type is defined:
+
+```ts
+export type ClassNames<ComponentName extends string, Slots> = {
+  [SlotName in keyof Slots]-?: SlotName extends 'root'
+    ? `fui-${ComponentName}`
+    : `fui-${ComponentName}-${Extract<SlotName, string>}`;
+};
+```
```

Comment:
_Safety check:_ that's cool 😍, but this does not work with TS 3.9 that we are targeting:

https://github.com/microsoft/fluentui/blob/08bb967395c61ccbf53a638b5e16f4b5723a03c9/apps/ts-minbar-test-react-components/src/index.ts#L10-L11

[TypeScript playground](https://www.typescriptlang.org/play?ts=3.9.7#code/KYDwDg9gTgLgBDAnmYcDCAbAhgZxwOSwFtgcAeNCIyAO2BpkJLlBnoBMc4cYoBLGgHMANHADKGCDBwA+OAF44AbwBQcOAG0JUpqgFwA1sEQQAZuMnSAugFoA-AC4LO4qlYcuAcigQpntepwdnAABqYArnw2ACRKlNQQdAy6AL4hAepOYZExcVS09IyuKbkAoiC8WADGMGTaRSSiPPxCMmkA3Cop7UA)

Reply:
😢 oh well. But at least we can ensure that every slot is specified, and then check the name formatting using a conformance test:
```ts
export type ClassNames<Slots> = {
  [SlotName in keyof Slots]-?: string;
};
```